### DISCUSSION TITLE
Including whitespace-aware blocks

### DISCUSSION THREAD

**Discussion by amenyashev** [eclipse-langium/langium#663] [Ideas]
Hello
you will including whitespace-aware blocks in langium? Example in xtext this - using synthetic tokens in the grammar by writing terminals of the form :'synthetic:<terminal name>'
Thank you

**Comment by msujew** [comment]
Hey @amenyashev,

we already allow to create synthetic tokens in Langium. You can just define the terminal in any way (something like `terminal INDENT: 'hello';`) and overwrite its exact behavior in the [`TokenBuilder`](https://github.com/langium/langium/blob/main/packages/langium/src/parser/token-builder.ts) service. We should probably write some documentation on that...

**Reply by amenyashev** [reply]
Thank you

**Reply by aabounegm** [reply]
Hi @msujew, I was wondering, has any documentation been written about this issue yet?

**Reply by msujew** [reply]
@aabounegm kind of, see https://github.com/eclipse-langium/langium-website/pull/132. I never got back to finish that one.

**Reply by aabounegm** [reply]
Update: the [IndentationAwareTokenBuilder](https://eclipse-langium.github.io/langium/classes/langium.IndentationAwareTokenBuilder.html) has been released in v3.2.0, and there is now [a recipe](https://langium.org/docs/recipes/lexing/indentation-sensitive-languages/) explaining how to use it.
